dbms/src/DataStreams/WindowBlockInputStream.cpp

```diff
@@ -272,12 +274,120 @@ Int64 WindowTransformAction::getPartitionEndRow(size_t block_rows)
     return left;
 }
 
+RowNumber WindowTransformAction::stepForward(const RowNumber & current_row, Int64 n)
+{
+    auto dist = distance(current_row, partition_start);
+    assert(dist >= 0);
+    if (dist <= n)
+        return partition_start;
+
+    RowNumber result_row = current_row;
+
+    // The step happens only in a block
+    if (static_cast<Int64>(result_row.row) >= n)
+    {
+        result_row.row -= n;
+        return result_row;
+    }
+
+    // The step happens between blocks
+    n -= result_row.row + 1;
+    --result_row.block;
+    result_row.row = blockAt(result_row).rows - 1;
+    while (n > 0)
+    {
+        auto & block = blockAt(result_row);
+        if (static_cast<Int64>(block.rows) > n)
+        {
+            result_row.row = block.rows - n - 1; // index, so we need to -1
+            break;
+        }
+        n -= block.rows;
+        --result_row.block;
+        result_row.row = blockAt(result_row).rows - 1;
+    }
+    return result_row;
+}
+
+std::tuple<RowNumber, bool> WindowTransformAction::stepBackward(const RowNumber & current_row, Int64 n)
+{
+    // Range of rows is [frame_start, frame_end),
+    // and frame_end position is behind the position of the last frame row.
+    // So we need to ++n.
+    ++n;
+
+    auto dist = distance(partition_end, current_row);
+    assert(dist >= 1);
```

Comment:
Is it possible for `partition_end` and `current_row` to be equal?

Reply:
> Is it possible for `partition_end` and `current_row` to be equal?

If the programme reaches to here, `partition_end` and `current_row` is impossible to be equal.